gapic/src/main/com/google/gapid/views/PipelineView.java

```diff
@@ -184,202 +192,293 @@ protected void onUiThreadError(Loadable.Message error) {
-      for (API.Pipeline result : pipelines) {
-        List<API.Stage> stages = result.getStagesList();
+      Composite stripComposite = withLayoutData( createComposite(folder, new FillLayout(SWT.HORIZONTAL)),
+          new GridData(SWT.FILL, SWT.TOP, true, false));
+      Label separator = withLayoutData( new Label(folder, SWT.SEPARATOR | SWT.HORIZONTAL),
+          new GridData(SWT.FILL, SWT.TOP, true, false));
+
+      StackLayout stageStack = new StackLayout();
+      Composite stageComposite = withLayoutData( createComposite(folder, stageStack),
+          new GridData(SWT.FILL, SWT.FILL, true, true));
+
+      for (int pipeIndex = 0; pipeIndex < pipelines.size(); pipeIndex++) {
+        List<API.Stage> stages = pipelines.get(pipeIndex).getStagesList();
+
+        for (int stageIndex = 0; stageIndex < stages.size(); stageIndex++) {
+          API.Stage stage = stages.get(stageIndex);
+          stageMap.put(stage.getDebugName(), createStage(stageComposite, stage));
+
+          Button stageButton = new Button(stripComposite, SWT.FLAT);
+          stageButton.setText(stage.getDebugName());
+          stageButton.addListener(SWT.Selection, e -> {
+            stageStack.topControl = stageMap.get(stage.getDebugName());
+            stageComposite.layout();
+            selectedStage = stage.getDebugName();
+            stripComposite.redraw();
+          });
+
+          stageButton.setBackground(getDisplay().getSystemColor(SWT.COLOR_LIST_BACKGROUND));
+          stageButton.setFont(theme.bigBoldFont());
+          stageButton.setToolTipText(stage.getStageName() + " (" + stage.getDebugName() + ")");
 
-        for (API.Stage stage : stages) {
-          TabItem item = createStandardTabItem(folder, stage.getDebugName());
+          if (!stage.getEnabled()) {
+            stageButton.setEnabled(false);
+            stageButton.addListener(SWT.Paint, e -> {
+              Rectangle areaSize = stageButton.getBounds();
+              e.gc.setBackground(getDisplay().getSystemColor(SWT.COLOR_LIST_FOREGROUND));
+              e.gc.setLineWidth((int)(areaSize.height * 0.1));
+              e.gc.drawLine(0, areaSize.height / 2, areaSize.width, areaSize.height / 2);
+              e.gc.drawPolygon(new int[] {0, 0, areaSize.width - 1, 0, areaSize.width - 1, areaSize.height - 1, 0, areaSize.height - 1});
+            });
+          } else {
+            stageButton.addListener(SWT.Paint, e -> {
+              Rectangle areaSize = stageButton.getBounds();
+              e.gc.setLineWidth((int)(areaSize.height * 0.1));
+
+              if (hoveredButton == stageButton) {
+                RGB hoverColor = getDisplay().getSystemColor(SWT.COLOR_BLUE).getRGB();
+                stageButton.setBackground(new Color(stageButton.getDisplay(), hoverColor.red+30, hoverColor.green+30, hoverColor.blue, 63));
+              } else {
+                stageButton.setBackground(getDisplay().getSystemColor(SWT.COLOR_LIST_BACKGROUND));
+              }
```

Comment:
Delete these six lines here. It is too late to change the button background color in the paint event (which happens after the widget has been drawn).
Further it is not good to create a new system resource (color) each time, and not to dispose them.